Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs

```diff
@@ -741,35 +719,38 @@ private static void CreateFakeFacilitiesAndSlots(IDbConnection db)
         {
             var occurrences = Enumerable.Range(10, OpportunityCount * 10)
-            .Select(n => new {
-                Id = n,
-                StartDate = Faker.Date.Soon(10).Truncate(TimeSpan.FromSeconds(1)),
-                TotalSpaces = Faker.Random.Bool() ? Faker.Random.Int(0, 50) : Faker.Random.Int(0, 3)
-            })
-            .Select(x => new OccurrenceTable
-            {
-                ClassId = decimal.ToInt32(x.Id / 10M),
-                Id = x.Id,
-                Deleted = false,
-                Start = x.StartDate,
-                End = x.StartDate + TimeSpan.FromMinutes(Faker.Random.Int(30, 360)),
-                TotalSpaces = x.TotalSpaces,
-                RemainingSpaces = x.TotalSpaces
-            })
-            .ToList();
+                .Select(id => new
+                {
+                    Id = id,
+                    StartDate = Faker.Date.Soon(10).Truncate(TimeSpan.FromSeconds(1)),
+                    TotalSpaces = Faker.Random.Bool() ? Faker.Random.Int(0, 50) : Faker.Random.Int(0, 3)
+                })
+                .Select(t => new OccurrenceTable
+                {
+                    ClassId = decimal.ToInt32(t.Id / 10M),
+                    Id = t.Id,
+                    Deleted = false,
+                    Start = t.StartDate,
+                    End = t.StartDate + TimeSpan.FromMinutes(Faker.Random.Int(30, 360)),
+                    TotalSpaces = t.TotalSpaces,
+                    RemainingSpaces = t.TotalSpaces
+                })
+                .ToList();
 
             var classes = Enumerable.Range(1, OpportunityCount)
-            .Select(id => new ClassTable
-            {
-                Id = id,
-                Deleted = false,
-                Title = $"{Faker.Commerce.ProductMaterial()} {Faker.PickRandomParam("Yoga", "Zumba", "Walking", "Cycling", "Running", "Jumping")}",
-                Price = decimal.Parse(Faker.Random.Bool() ? "0.00" : Faker.Commerce.Price(0, 20)),
-                RequiresApproval = Faker.Random.Bool(),
-                SellerId = Faker.Random.Long(1, 3),
-                ValidFromBeforeStartDate = ValidFromBeforeStartDate(Faker.Random)
-            })
-            .ToList();
+                .Select(id => new ClassTable
+                {
+                    Id = id,
+                    Deleted = false,
+                    Title = $"{Faker.Commerce.ProductMaterial()} {Faker.PickRandomParam("Yoga", "Zumba", "Walking", "Cycling", "Running", "Jumping")}",
+                    Price = decimal.Parse(Faker.Random.Bool() ? "0.00" : Faker.Commerce.Price(0, 20)),
+                    RequiresApproval = Faker.Random.Bool(),
+                    SellerId = Faker.Random.Long(1, 3),
+                    ValidFromBeforeStartDate = ValidFromBeforeStartDate(
+                        Faker.Random.Bool() ? Faker.Random.Bool() : (bool?)null,
+                        occurrences.Single(o => o.Id == id * 10).Start)
```

Comment:
Is there a more robust way of achieving this? Would be good to add a comment on what this logic does too

Reply:
It's robust, as it mirrors the logic above. I could make it more efficient by using a dictionary. I wondered whether to bother, but now you've convinced me.